README.md
# GTM: Google Toolbox for Mac #

**Project site** <https://github.com/google/google-toolbox-for-mac><br>
**Discussion group** <http://groups.google.com/group/google-toolbox-for-mac>


## Release History ##

Release 2.0.??
Changes since 1.6.0
??-??-??

- Removed iPhone/GTMABAddressBook in favor of AddressBook/GTMABAddressBook.

- Removed Foundation/GTMHTTPServer and UnitTesting/GTMTestHTTPServer, they
  are going to go live with the fetcher used by GData (since they were done
  for that testing).

- Removed Foundation/GTMBase64 and Foundation/GTMNSData+Hex in favor of
  Foundation/GTMStringEncoding.

- Added Foundation/GTMURITemplate to support the pending standard.

- Changed the xcconfig files so that the SDK and the minimum supported OS
  version must be set in the project file.  This is the model Apple is pushing
  for and since they are removing older SDKs with each tool chain release it
  has forced GTM into this model. By default the SDK will be set to the most
  recent SDK installed.

- Initial support for using the Xcode provided OCUnit/XCTest for unittesting on iOS.
  Define GTM_IPHONE_USE_SENTEST or GTM_USING_XCTEST to 1 to use this.
  GTM_IPHONE_USE_SENTEST is going to be deprecated shortly.

- Removed support for Garbage Collection, leaving just the shell for other code
  that might have depended on some of the constants/method GTM provided.

- Removed GTMNSNumber+64Bit methods as obsolete.


Release 1.6.0
Changes since 1.5.1
18-August-2010

- Added GTMNSImage+SearchCache for fetching images based on a variety of
  specification methods (path, OSType, etc)

- Added GTMFadeTruncatingTextFieldCell, for eliding with a gradient

- Added GTMWindowSheetController for creating and controlling tab-modal sheets.

- Added GTMNSArray+Merge for merging one array into another with or without
  a custom merging function, returning a new array with the merged contents.

- Added GTMSignalHandler for simple signal handling (via kqueue/runloop). This
  has gotten an api tweak, so some code that started using it will need
  updating.  Initial landing had a bug where it could leak memory due to
  how CFRunLoops work, now fixed.

- Fixed up GTMIPhoneUnitTestDelegate to be pickier about which tests it runs

- Added GTMNSString+URLArguments to GTMiPhone

- Added GTMHTTPFetcher and GTMHTTPServer to GTMiPhone

- Made sure that build would work with iPhone device attached, and that all
  tests run directly on the phone.

- Added GTMValidatingContainers which are a set of mutable container classes
  that allow you to have a selector on a target that is called to verify that
  the objects being put into the container are valid. This can be controlled
  at compile time so that you don't take the performance hit in a release build.

- Added GTMPath, which represents an existing absolute path on the file system.
  It also makes it very easy to construct new paths in the file system as well
  as whole directory hierarchies.

- Added GTMNSString+Replace for a common replacement need.

- Added NSString+FindFolder for two comment helpers for building paths to common
  locations.

- Added GTMLargeTypeWindow for doing display windows similar to Address Book
  Large Type display for phone numbers.

- Removed GTMNSWorkspace+ScreenSaver as it has always been a little dodgy due
  to it's dependencies on undocumented frameworks, and the ScreenSaver
  framework doesn't play nicely in GC mode.

- Added property methods to GTMHTTPFetcher.  These are convenient alternatives
  to storing an NSDictionary in the userData.

- Renamed GTMDevLog.m to GTMDevLogUnitTestingBridge.m and added some more
  comments where it comes into play to hopefully make it more clear that it
  isn't needed in most cases.

- Fixed a potential GTMHTTPFetcher crash on failed authentication.

- Added a obj-c logging package, GTMLogger, for applications that want an
  application level logging system.  See GTMLogger.h, GTMLogger+ASL.h, and
  GTMLoggerRingBufferWriter.h for what the basic system and two optional
  additions can do.

- Added GTMNSMakeUncollectable for forcing objects to survive in a GC world.

- Added GTMCFAutorelease to make the [GTMNSMakeCollectable(cfFoo) autorelease]
  simpler and clearer, it's now just GTMCFAutorelease(cfFoo), and works in
  both GC and non-GC world.

- Added GTMIsGarbageCollectionEnabled to GTMGarbageCollection.h.  See the note
  there for it's usage.

- Disabled the unittests for things on top of NSAppleScript in a GC world since
  Apple has bugs and it can crash.  See the unittest for a note about it.

- GTMStackTrace now can figure out ObjC symbols. Downside it is now ObjC only.

- GTMFourCharCode can now be used with NSAppleEventDescriptors easily.
  typeType, typeKeyword, typeApplSignature, and typeEnumerated all get
  turned into GTMFourCharCodes.

- Fixed up crash in GTMLoggerRingBufferWriter when used with GC on.

- Significant updates to GTMNSAppleScript+Handler allowing you to
  list all handlers and properties (including inherited) and cleans up
  several errors in how scripting was being handled.

- Added GTMGetURLHandler class that gives you a very easy way of supporting
  Get URL events just by adding a key to your plists, and adding a single
  method to your class. See GTMGetURLHandler.m for more details.

- Added XcodeProject, AppleScript, and InterfaceBuilder Spotlight Plugins.
  Allows you to index .xcodeproj, .scpt, .scptd, .xib, .nib, and
  .aib files. See ReadMes beside individual projects in SpotlightPlugins.

- Added GTMExceptionalInlines for dealing with cases where you get
  warning: variable 'r' might be clobbered by 'longjmp' or 'vfork'
  when using certain Apple inlined functions in @synchronized/@try blocks.

- Updated to Xcode 3.1 so the GTM and iPhone project have the same baseline.
  The code should work in other version of xcode, but the projects and
  xcconfig files now use 3.1 features.

- Added GTMABAddressBook which is a cocoa wrapper for the 'C' AddressBook
  APIs on the iPhone.

- Added several set environment variable statements to RunIPhoneUnitTest.sh
  to encourage bugs to come out of the woodwork.

- Added GTMTestTimer.h for doing high fidelity timings.

- Added ability to control using zombies to iPhone unittest script. It can be
  controlled by the GTM_DISABLE_ZOMBIES environment variable

- Added ability to control termination to iPhone unittest script. It can be
  controlled by the GTM_DISABLE_TERMINATION environment variable

- Fixed several leaks found with leak checking enabled.

- Updated the iPhone xcconfigs to support the different OS versions.

- GTM_INLINE will make sure a function gets inlined, and provides a consistent
  way for all GTM code to do it.

- Added GTMDebugThreadValidation to allow you to enforce the fact that your
  code must run in the main thread in DEBUG builds.

- Updated some internals of the iPhone unittesting so it doesn't double print
  the test descriptions, file names, or lines numbers of a test failure line.
  Also includes the test names in the error output.

- Changed the xcconfigs so that know it's easier to set different settings at
  the different levels and not accidentally overwrite settings set at lower
  levels in the "settings collapse". Also tightened up warnings significantly.

- Changed how gtm_unitTestExposedBindingsTestValues works. If you have an
  implementation of gtm_unitTestExposedBindingsTestValues in your own code
  you will need to update to the new way of calling. See implementations in
  GTMNSObject+BindingUnitTesting.m for details.

- Added support for grabbing the build number for a particular OS in
  GTMSystemVersion and easily comparing it to known build numbers, and switched
  some types from in GTMSystemVersion from "int" to SInt32 to make 64 bit work
  better.

- Added support for SnowLeopard (10A96). We build cleanly with the 10.6 SDKs and
  all radar checks were updated accordingly. Build All script was also updated
  to build on SnowLeopard if you have the SDK available.

- Turned off building ppc64 GTM because the SnowLeopard SDK currently
  doesn't have ppc64 support, so SenTestCase isn't defined. This makes it
  impossible to build the ppc64 10.5 config on SnowLeopard. We have left the
  setting in the xcconfig for those of you who need it, but have disabled
  it in the GTM project settings.

- Turned on stack smashing protection on the debug builds for all Leopard
  and above.

- Added ability to easily do leak checking by defining the GTM_ENABLE_LEAKS
  environment variable. It isn't on by default because several of Apple's
  frameworks leak. You can work around these false positives by using the
  GTM_LEAKS_SYMBOLS_TO_IGNORE environment variable. Also if you turn on leaks
  make sure to turn off zombies by defining the GTM_DISABLE_ZOMBIES variable,
  otherwise every memory allocation you do will look like a leak.

- Added has ability to check if a script has an open handler to
  GTMNSAppleScript+Handler.

- GTMStackTrace support for building a trace from the call stack in an
  NSException (for 10.5+ and iPhone).

- GTMStackTrace works on 10.5+ (and iPhone) using NSThread to build the call
  stack.

- GTMLightweightProxy for breaking retain cycles.

- Added GTM_EXTERN that makes it easier to mix and match objc and objc++ code.

- Added GTMHotKeysTextField for display and editing of hot key settings.

- Added GTMCarbonEvent for dealing with Carbon Events and HotKeys in a ObjC
  like way.

- Backported the Atomic Barrier Swap functions for Objective C back to Tiger.

- Added a variety of new functions to GTMUnitTestingUtilities for checking
  if the screensaver is in the way, waiting on user events, and generating
  keystrokes.

- If you are using any Carbon routines that log (DebugStr, AssertMacros.h) and
  use GTMUnitTestDevLog, the log routines now go through _GTMDevLog so that
  they can be caught in GTMUnitTestDevLog and verified like any _GTMDevLog calls
  you may make. For an example of this in action see GTMCarbonEventTest.m.
  Since we have turned this on, we have turned off using _debug frameworks
  from the RunUnitTests.sh because it was reporting a pile of uninteresting
  issues that were interfering with unittests.

- Added GTMFileSystemKQueue.  It provides a simple wrapper for kqueuing
  something in the file system and tracking changes to it. Initial landing
  had a bug where it could leak memory due to how CFRunLoops work, now fixed.

- RunIPhoneUnitTest.sh now cleans up the user home directory and creates

ReleaseNotes.md
# GTM: Google Toolbox for Mac #
## Release History ##

*NOTE:* There haven't been release as most users of this code have just been
pulling trunk the HEAD code to get the latest updates/fixes.  This code is also
a little different in that it is more of a "collection" of things than a single
library providing one thing, so by pulling in individual source files,
developers can pick up the individual parts they need.

**Release 2.0.??**<br>
Changes since 1.6.0<br>
??-??-??

- Removed iPhone/GTMABAddressBook in favor of AddressBook/GTMABAddressBook.

- Removed Foundation/GTMHTTPServer and UnitTesting/GTMTestHTTPServer, they
  are going to go live with the fetcher used by GData (since they were done
  for that testing).

- Removed Foundation/GTMBase64 and Foundation/GTMNSData+Hex in favor of
  Foundation/GTMStringEncoding.

- Added Foundation/GTMURITemplate to support the pending standard.

- Changed the xcconfig files so that the SDK and the minimum supported OS
  version must be set in the project file.  This is the model Apple is pushing
  for and since they are removing older SDKs with each tool chain release it
  has forced GTM into this model. By default the SDK will be set to the most
  recent SDK installed.

- Initial support for using the Xcode provided XCTest for unittesting on iOS.
  Define GTM_USING_XCTEST to 1 to use this.

- Removed support for Garbage Collection, leaving just the shell for other code
  that might have depended on some of the constants/method GTM provided.

- Removed GTMNSNumber+64Bit methods as obsolete.

- Removed GTM_ENABLE_LEAKS.

- Removed GTMGarbageCollection.h.


**Release 1.6.0**<br>
Changes since 1.5.1<br>
18-August-2010

- Added GTMNSImage+SearchCache for fetching images based on a variety of
  specification methods (path, OSType, etc)

- Added GTMFadeTruncatingTextFieldCell, for eliding with a gradient

- Added GTMWindowSheetController for creating and controlling tab-modal sheets.

- Added GTMNSArray+Merge for merging one array into another with or without
  a custom merging function, returning a new array with the merged contents.

- Added GTMSignalHandler for simple signal handling (via kqueue/runloop). This
  has gotten an api tweak, so some code that started using it will need
  updating.  Initial landing had a bug where it could leak memory due to
  how CFRunLoops work, now fixed.

- Fixed up GTMIPhoneUnitTestDelegate to be pickier about which tests it runs

- Added GTMNSString+URLArguments to GTMiPhone

- Added GTMHTTPFetcher and GTMHTTPServer to GTMiPhone

- Made sure that build would work with iPhone device attached, and that all
  tests run directly on the phone.

- Added GTMValidatingContainers which are a set of mutable container classes
  that allow you to have a selector on a target that is called to verify that
  the objects being put into the container are valid. This can be controlled
  at compile time so that you don't take the performance hit in a release build.

- Added GTMPath, which represents an existing absolute path on the file system.
  It also makes it very easy to construct new paths in the file system as well
  as whole directory hierarchies.

- Added GTMNSString+Replace for a common replacement need.

- Added NSString+FindFolder for two comment helpers for building paths to common
  locations.

- Added GTMLargeTypeWindow for doing display windows similar to Address Book
  Large Type display for phone numbers.

- Removed GTMNSWorkspace+ScreenSaver as it has always been a little dodgy due
  to it's dependencies on undocumented frameworks, and the ScreenSaver
  framework doesn't play nicely in GC mode.

- Added property methods to GTMHTTPFetcher.  These are convenient alternatives
  to storing an NSDictionary in the userData.

- Renamed GTMDevLog.m to GTMDevLogUnitTestingBridge.m and added some more
  comments where it comes into play to hopefully make it more clear that it
  isn't needed in most cases.

- Fixed a potential GTMHTTPFetcher crash on failed authentication.

- Added a obj-c logging package, GTMLogger, for applications that want an
  application level logging system.  See GTMLogger.h, GTMLogger+ASL.h, and
  GTMLoggerRingBufferWriter.h for what the basic system and two optional
  additions can do.

- Added GTMNSMakeUncollectable for forcing objects to survive in a GC world.

- Added GTMCFAutorelease to make the [GTMNSMakeCollectable(cfFoo) autorelease]
  simpler and clearer, it's now just GTMCFAutorelease(cfFoo), and works in
  both GC and non-GC world.

- Added GTMIsGarbageCollectionEnabled to GTMGarbageCollection.h.  See the note
  there for it's usage.

- Disabled the unittests for things on top of NSAppleScript in a GC world since
  Apple has bugs and it can crash.  See the unittest for a note about it.

- GTMStackTrace now can figure out ObjC symbols. Downside it is now ObjC only.

- GTMFourCharCode can now be used with NSAppleEventDescriptors easily.
  typeType, typeKeyword, typeApplSignature, and typeEnumerated all get
  turned into GTMFourCharCodes.

- Fixed up crash in GTMLoggerRingBufferWriter when used with GC on.

- Significant updates to GTMNSAppleScript+Handler allowing you to

GTMDefines.h
//
// GTMDefines.h
//
//  Copyright 2008 Google Inc.
//
//  Licensed under the Apache License, Version 2.0 (the "License"); you may not
//  use this file except in compliance with the License.  You may obtain a copy
//  of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
//  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
//  License for the specific language governing permissions and limitations under
//  the License.
//

// ============================================================================

#ifdef __APPLE__
# include <AvailabilityMacros.h>
# include <TargetConditionals.h>
#endif

#ifdef __OBJC__
# ifndef __MULLE_OBJC_RUNTIME__
#  include <Foundation/NSObjCRuntime.h>
# endif
#endif  // __OBJC__

#if TARGET_OS_IPHONE
#include <Availability.h>
#endif  // TARGET_OS_IPHONE

// Not all MAC_OS_X_VERSION_10_X macros defined in past SDKs
#ifndef MAC_OS_X_VERSION_10_5
  #define MAC_OS_X_VERSION_10_5 1050
#endif
#ifndef MAC_OS_X_VERSION_10_6
  #define MAC_OS_X_VERSION_10_6 1060
#endif
#ifndef MAC_OS_X_VERSION_10_7
  #define MAC_OS_X_VERSION_10_7 1070
#endif
#ifndef MAC_OS_X_VERSION_10_8
  #define MAC_OS_X_VERSION_10_8 1080
#endif
#ifndef MAC_OS_X_VERSION_10_9
  #define MAC_OS_X_VERSION_10_9 1090
#endif

// Not all __IPHONE_X macros defined in past SDKs
#ifndef __IPHONE_3_0
  #define __IPHONE_3_0 30000
#endif
#ifndef __IPHONE_3_1
  #define __IPHONE_3_1 30100
#endif
#ifndef __IPHONE_3_2
  #define __IPHONE_3_2 30200
#endif
#ifndef __IPHONE_4_0
  #define __IPHONE_4_0 40000
#endif
#ifndef __IPHONE_4_3
  #define __IPHONE_4_3 40300
#endif
#ifndef __IPHONE_5_0
  #define __IPHONE_5_0 50000
#endif
#ifndef __IPHONE_5_1
  #define __IPHONE_5_1 50100
#endif
#ifndef __IPHONE_6_0
  #define __IPHONE_6_0 60000
#endif
#ifndef __IPHONE_6_1
  #define __IPHONE_6_1 60100
#endif
#ifndef __IPHONE_7_0
  #define __IPHONE_7_0 70000
#endif
#ifndef __IPHONE_7_1
  #define __IPHONE_7_1 70100
#endif

// ----------------------------------------------------------------------------
// CPP symbols that can be overridden in a prefix to control how the toolbox
// is compiled.
// ----------------------------------------------------------------------------


// By setting the GTM_CONTAINERS_VALIDATION_FAILED_LOG and
// GTM_CONTAINERS_VALIDATION_FAILED_ASSERT macros you can control what happens
// when a validation fails. If you implement your own validators, you may want
// to control their internals using the same macros for consistency.
#ifndef GTM_CONTAINERS_VALIDATION_FAILED_ASSERT
  #define GTM_CONTAINERS_VALIDATION_FAILED_ASSERT 0
#endif

// Ensure __has_feature and __has_extension are safe to use.
// See http://clang-analyzer.llvm.org/annotations.html
#ifndef __has_feature      // Optional.
  #define __has_feature(x) 0 // Compatibility with non-clang compilers.
#endif

#ifndef __has_extension
  #define __has_extension __has_feature // Compatibility with pre-3.0 compilers.
#endif

// Give ourselves a consistent way to do inlines.  Apple's macros even use
// a few different actual definitions, so we're based off of the foundation
// one.
#if !defined(GTM_INLINE)
  #if (defined (__GNUC__) && (__GNUC__ == 4)) || defined (__clang__)
    #define GTM_INLINE static __inline__ __attribute__((always_inline))
  #else
    #define GTM_INLINE static __inline__
  #endif
#endif

// Give ourselves a consistent way of doing externs that links up nicely
// when mixing objc and objc++
#if !defined (GTM_EXTERN)
  #if defined __cplusplus
    #define GTM_EXTERN extern "C"
    #define GTM_EXTERN_C_BEGIN extern "C" {
    #define GTM_EXTERN_C_END }
  #else
    #define GTM_EXTERN extern
    #define GTM_EXTERN_C_BEGIN
    #define GTM_EXTERN_C_END
  #endif
#endif

// Give ourselves a consistent way of exporting things if we have visibility
// set to hidden.
#if !defined (GTM_EXPORT)
  #define GTM_EXPORT __attribute__((visibility("default")))
#endif

// Give ourselves a consistent way of declaring something as unused. This
// doesn't use __unused because that is only supported in gcc 4.2 and greater.
#if !defined (GTM_UNUSED)
#define GTM_UNUSED(x) ((void)(x))
#endif

// _GTMDevLog & _GTMDevAssert
//
// _GTMDevLog & _GTMDevAssert are meant to be a very lightweight shell for
// developer level errors.  This implementation simply macros to NSLog/NSAssert.
// It is not intended to be a general logging/reporting system.
//
// Please see http://code.google.com/p/google-toolbox-for-mac/wiki/DevLogNAssert
// for a little more background on the usage of these macros.
//
//    _GTMDevLog           log some error/problem in debug builds
//    _GTMDevAssert        assert if condition isn't met w/in a method/function
//                           in all builds.
//
// To replace this system, just provide different macro definitions in your
// prefix header.  Remember, any implementation you provide *must* be thread
// safe since this could be called by anything in what ever situtation it has
// been placed in.
//

// We only define the simple macros if nothing else has defined this.
#ifndef _GTMDevLog

#ifdef DEBUG
  #define _GTMDevLog(...) NSLog(__VA_ARGS__)
#else
  #define _GTMDevLog(...) do { } while (0)
#endif

#endif // _GTMDevLog

#ifndef _GTMDevAssert
// we directly invoke the NSAssert handler so we can pass on the varargs
// (NSAssert doesn't have a macro we can use that takes varargs)
#if !defined(NS_BLOCK_ASSERTIONS)
  #define _GTMDevAssert(condition, ...)                                       \
    do {                                                                      \
      if (!(condition)) {                                                     \
        [[NSAssertionHandler currentHandler]                                  \
            handleFailureInFunction:(NSString *)                              \
                                        [NSString stringWithUTF8String:__PRETTY_FUNCTION__] \
                               file:(NSString *)[NSString stringWithUTF8String:__FILE__]  \
                         lineNumber:__LINE__                                  \
                        description:__VA_ARGS__];                             \
      }                                                                       \
    } while(0)
#else // !defined(NS_BLOCK_ASSERTIONS)
  #define _GTMDevAssert(condition, ...) do { } while (0)
#endif // !defined(NS_BLOCK_ASSERTIONS)

#endif // _GTMDevAssert

// _GTMCompileAssert
//
// Note:  Software for current compilers should just use _Static_assert directly
// instead of this macro.
//
// _GTMCompileAssert is an assert that is meant to fire at compile time if you
// want to check things at compile instead of runtime. For example if you
// want to check that a wchar is 4 bytes instead of 2 you would use
// _GTMCompileAssert(sizeof(wchar_t) == 4, wchar_t_is_4_bytes_on_OS_X)
// Note that the second "arg" is not in quotes, and must be a valid processor
// symbol in it's own right (no spaces, punctuation etc).

// Wrapping this in an #ifndef allows external groups to define their own
// compile time assert scheme.
#ifndef _GTMCompileAssert
  #if __has_feature(c_static_assert) || __has_extension(c_static_assert)
    #define _GTMCompileAssert(test, msg) _Static_assert((test), #msg)
  #else
    // Pre-Xcode 7 support.
    //
    // We got this technique from here:
    // http://unixjunkie.blogspot.com/2007/10/better-compile-time-asserts_29.html
    #define _GTMCompileAssertSymbolInner(line, msg) _GTMCOMPILEASSERT ## line ## __ ## msg
    #define _GTMCompileAssertSymbol(line, msg) _GTMCompileAssertSymbolInner(line, msg)
    #define _GTMCompileAssert(test, msg) \
      typedef char _GTMCompileAssertSymbol(__LINE__, msg) [ ((test) ? 1 : -1) ]
  #endif  // __has_feature(c_static_assert) || __has_extension(c_static_assert)
#endif // _GTMCompileAssert

// ----------------------------------------------------------------------------
// CPP symbols defined based on the project settings so the GTM code has
// simple things to test against w/o scattering the knowledge of project
// setting through all the code.
// ----------------------------------------------------------------------------

// Provide a single constant CPP symbol that all of GTM uses for ifdefing
// iPhone code.
#if TARGET_OS_IPHONE // iPhone SDK
  // For iPhone specific stuff
  #define GTM_IPHONE_SDK 1
  #if TARGET_IPHONE_SIMULATOR
    #define GTM_IPHONE_DEVICE 0
    #define GTM_IPHONE_SIMULATOR 1
  #else
    #define GTM_IPHONE_DEVICE 1
    #define GTM_IPHONE_SIMULATOR 0
  #endif  // TARGET_IPHONE_SIMULATOR
  // By default, GTM has provided it's own unittesting support, define this
  // to use the support provided by Xcode, especially for the Xcode4 support
  // for unittesting.
  // This is going to be deprecated as Apple is deprecating SenTest.
  #ifndef GTM_IPHONE_USE_SENTEST
    #define GTM_IPHONE_USE_SENTEST 0
  #endif
  // Define this to use XCTest instead of OCUnit/SenTest.
  #ifndef GTM_USING_XCTEST
    #define GTM_USING_XCTEST 0
  #endif
  #if GTM_IPHONE_USE_SENTEST && GTM_USING_XCTEST
    #error Can't define both GTM_IPHONE_USE_SENTEST and GTM_USING_XCTEST
  #endif
  #define GTM_MACOS_SDK 0
#else
  // For MacOS specific stuff
  #define GTM_MACOS_SDK 1
  #define GTM_IPHONE_SDK 0
  #define GTM_IPHONE_SIMULATOR 0
  #define GTM_IPHONE_DEVICE 0
  #define GTM_IPHONE_USE_SENTEST 0
  #define GTM_USING_XCTEST 0
#endif

// Some of our own availability macros
#if GTM_MACOS_SDK
#define GTM_AVAILABLE_ONLY_ON_IPHONE UNAVAILABLE_ATTRIBUTE
#define GTM_AVAILABLE_ONLY_ON_MACOS
#else
#define GTM_AVAILABLE_ONLY_ON_IPHONE
#define GTM_AVAILABLE_ONLY_ON_MACOS UNAVAILABLE_ATTRIBUTE
#endif

// GC was dropped by Apple, define the old constant incase anyone still keys
// off of it.
#ifndef GTM_SUPPORT_GC
  #define GTM_SUPPORT_GC 0
#endif

// To simplify support for 64bit (and Leopard in general), we provide the type
// defines for non Leopard SDKs
#if !(MAC_OS_X_VERSION_MAX_ALLOWED >= MAC_OS_X_VERSION_10_5)
 // NSInteger/NSUInteger and Max/Mins
  #ifndef NSINTEGER_DEFINED
    #if (defined(__LP64__) && __LP64__) || NS_BUILD_32_LIKE_64
      typedef long NSInteger;
      typedef unsigned long NSUInteger;
    #else
      typedef int NSInteger;
      typedef unsigned int NSUInteger;
    #endif
    #define NSIntegerMax    LONG_MAX
    #define NSIntegerMin    LONG_MIN
    #define NSUIntegerMax   ULONG_MAX
    #define NSINTEGER_DEFINED 1
  #endif  // NSINTEGER_DEFINED
  // CGFloat
  #ifndef CGFLOAT_DEFINED
    #if defined(__LP64__) && __LP64__
      // This really is an untested path (64bit on Tiger?)
      typedef double CGFloat;
      #define CGFLOAT_MIN DBL_MIN
      #define CGFLOAT_MAX DBL_MAX
      #define CGFLOAT_IS_DOUBLE 1
    #else /* !defined(__LP64__) || !__LP64__ */
      typedef float CGFloat;
      #define CGFLOAT_MIN FLT_MIN
      #define CGFLOAT_MAX FLT_MAX
      #define CGFLOAT_IS_DOUBLE 0
    #endif /* !defined(__LP64__) || !__LP64__ */
    #define CGFLOAT_DEFINED 1
  #endif // CGFLOAT_DEFINED
#endif  // MAC_OS_X_VERSION_MIN_REQUIRED < MAC_OS_X_VERSION_10_5

// Some support for advanced clang static analysis functionality
#ifndef NS_RETURNS_RETAINED
  #if __has_feature(attribute_ns_returns_retained)
    #define NS_RETURNS_RETAINED __attribute__((ns_returns_retained))
  #else
    #define NS_RETURNS_RETAINED
  #endif
#endif

#ifndef NS_RETURNS_NOT_RETAINED
  #if __has_feature(attribute_ns_returns_not_retained)
    #define NS_RETURNS_NOT_RETAINED __attribute__((ns_returns_not_retained))
  #else
    #define NS_RETURNS_NOT_RETAINED
  #endif
#endif

#ifndef CF_RETURNS_RETAINED
  #if __has_feature(attribute_cf_returns_retained)
    #define CF_RETURNS_RETAINED __attribute__((cf_returns_retained))
  #else
    #define CF_RETURNS_RETAINED
  #endif
#endif

#ifndef CF_RETURNS_NOT_RETAINED
  #if __has_feature(attribute_cf_returns_not_retained)
    #define CF_RETURNS_NOT_RETAINED __attribute__((cf_returns_not_retained))
  #else
    #define CF_RETURNS_NOT_RETAINED
  #endif
#endif

#ifndef NS_CONSUMED
  #if __has_feature(attribute_ns_consumed)
    #define NS_CONSUMED __attribute__((ns_consumed))
  #else
    #define NS_CONSUMED
  #endif
#endif

#ifndef CF_CONSUMED
  #if __has_feature(attribute_cf_consumed)
    #define CF_CONSUMED __attribute__((cf_consumed))
  #else
    #define CF_CONSUMED
  #endif
#endif

#ifndef NS_CONSUMES_SELF
  #if __has_feature(attribute_ns_consumes_self)
    #define NS_CONSUMES_SELF __attribute__((ns_consumes_self))
  #else
    #define NS_CONSUMES_SELF
  #endif
#endif

// Defined on 10.6 and above.
#ifndef NS_FORMAT_ARGUMENT
  #define NS_FORMAT_ARGUMENT(A)
#endif

// Defined on 10.6 and above.
#ifndef NS_FORMAT_FUNCTION
  #define NS_FORMAT_FUNCTION(F,A)
#endif

// Defined on 10.6 and above.
#ifndef CF_FORMAT_ARGUMENT
  #define CF_FORMAT_ARGUMENT(A)
#endif

// Defined on 10.6 and above.
#ifndef CF_FORMAT_FUNCTION
  #define CF_FORMAT_FUNCTION(F,A)
#endif

#ifndef GTM_NONNULL
  #if defined(__has_attribute)
    #if __has_attribute(nonnull)
      #define GTM_NONNULL(x) __attribute__((nonnull x))
    #else
      #define GTM_NONNULL(x)
    #endif
  #else
    #define GTM_NONNULL(x)
  #endif
#endif

// Invalidates the initializer from which it's called.
#ifndef GTMInvalidateInitializer
  #if __has_feature(objc_arc)
    #define GTMInvalidateInitializer() \
      do { \
        [self class]; /* Avoid warning of dead store to |self|. */ \
        _GTMDevAssert(NO, @"Invalid initializer."); \
        return nil; \
      } while (0)
  #else
    #define GTMInvalidateInitializer() \
      do { \
        [self release]; \
        _GTMDevAssert(NO, @"Invalid initializer."); \
        return nil; \
      } while (0)
  #endif
#endif

#ifndef GTMCFAutorelease
  // GTMCFAutorelease returns an id.  In contrast, Apple's CFAutorelease returns
  // a CFTypeRef.
  #if __has_feature(objc_arc)
    #define GTMCFAutorelease(x) CFBridgingRelease(x)
  #else
    #define GTMCFAutorelease(x) ([(id)x autorelease])
  #endif
#endif

#ifdef __OBJC__

// Declared here so that it can easily be used for logging tracking if
// necessary. See GTMUnitTestDevLog.h for details.
@class NSString;
GTM_EXTERN void _GTMUnitTestDevLog(NSString *format, ...) NS_FORMAT_FUNCTION(1, 2);

// Macro to allow you to create NSStrings out of other macros.
// #define FOO foo
// NSString *fooString = GTM_NSSTRINGIFY(FOO);
#if !defined (GTM_NSSTRINGIFY)
  #define GTM_NSSTRINGIFY_INNER(x) @#x
  #define GTM_NSSTRINGIFY(x) GTM_NSSTRINGIFY_INNER(x)
#endif

// Macro to allow fast enumeration when building for 10.5 or later, and
// reliance on NSEnumerator for 10.4.  Remember, NSDictionary w/ FastEnumeration
// does keys, so pick the right thing, nothing is done on the FastEnumeration
// side to be sure you're getting what you wanted.
#ifndef GTM_FOREACH_OBJECT
  #if TARGET_OS_IPHONE || !(MAC_OS_X_VERSION_MIN_REQUIRED < MAC_OS_X_VERSION_10_5)
    #define GTM_FOREACH_ENUMEREE(element, enumeration) \
      for (element in enumeration)
    #define GTM_FOREACH_OBJECT(element, collection) \
      for (element in collection)
    #define GTM_FOREACH_KEY(element, collection) \
      for (element in collection)
  #else
    #define GTM_FOREACH_ENUMEREE(element, enumeration) \
      for (NSEnumerator *_ ## element ## _enum = enumeration; \
           (element = [_ ## element ## _enum nextObject]) != nil; )
    #define GTM_FOREACH_OBJECT(element, collection) \
      GTM_FOREACH_ENUMEREE(element, [collection objectEnumerator])
    #define GTM_FOREACH_KEY(element, collection) \
      GTM_FOREACH_ENUMEREE(element, [collection keyEnumerator])
  #endif
#endif

// ============================================================================

// To simplify support for both Leopard and Snow Leopard we declare
// the Snow Leopard protocols that we need here.
#if !defined(GTM_10_6_PROTOCOLS_DEFINED) && !(MAC_OS_X_VERSION_MAX_ALLOWED >= MAC_OS_X_VERSION_10_6)
#define GTM_10_6_PROTOCOLS_DEFINED 1
@protocol NSConnectionDelegate
@end
@protocol NSAnimationDelegate
@end
@protocol NSImageDelegate
@end
@protocol NSTabViewDelegate
@end
#endif  // !defined(GTM_10_6_PROTOCOLS_DEFINED) && !(MAC_OS_X_VERSION_MAX_ALLOWED >= MAC_OS_X_VERSION_10_6)

// GTM_SEL_STRING is for specifying selector (usually property) names to KVC
// or KVO methods.
// In debug it will generate warnings for undeclared selectors if
// -Wunknown-selector is turned on.
// In release it will have no runtime overhead.
#ifndef GTM_SEL_STRING
  #ifdef DEBUG
    #define GTM_SEL_STRING(selName) NSStringFromSelector(@selector(selName))
  #else
    #define GTM_SEL_STRING(selName) @#selName
  #endif  // DEBUG
#endif  // GTM_SEL_STRING

#ifndef GTM_WEAK
  #if defined(__OBJC_GC__)
    // In -fobjc-gc mode, __weak means "a reference not visible to the gargabe
    // collector".  __weak references are set to zero when their pointee is
    // collected.  __weak is not needed to prevent cycles because cycles
    // are cleaned up fine by the garbage collector.
    #define GTM_WEAK __weak
  #elif __has_feature(objc_arc_weak)
    // With ARC enabled, __weak means a reference that isn't implicitly
    // retained.  __weak objects are accessed through runtime functions, so
    // they are zeroed out, but this requires OS X 10.7+.
    // At clang r251041+, ARC-style zeroing weak references even work in
    // non-ARC mode.
    #define GTM_WEAK __weak
  #elif __has_feature(objc_arc)
    // ARC, but targeting 10.6 or older, where zeroing weak references don't
    // exist.
    #define GTM_WEAK __unsafe_unretained
  #else
    // With manual reference counting, __weak used to be silently ignored.
    // clang r251041 gives it the ARC semantics instead.  This means they
    // now require a deployment target of 10.7, while some clients of GTM
    // still target 10.6.  In these cases, expand to __unsafe_unretained instead
    #define GTM_WEAK
  #endif
#endif

#endif  // __OBJC__
